### DISCUSSION TITLE
Hub to use instead of Boost, so that I can communicate with external devices

### DISCUSSION THREAD

**Discussion by tomsaffell** [pybricks/support#1247] [General]
Hi, 

I'm planning out my first project, which will communicate with an external device, and I just discovered that the Boost Move hub (of which I own 2, plus motors) doesn't support that (thanks for documenting that well). I plan on buying a different hub. Questions:

1) Do _all_ other hubs support it? (usys / uselect) All equally well / reliably?
2) Which other hubs are compatible with the motors that come in Boost sets? (I'm considering 88009-1 and 88012-1)

thanks

tom

**Comment by dlech** [comment]
Since firmware version v3.3.0b8 there is `pybricks.tools.read_input_byte()` that can be used instead. It is available on all hubs, including the Move hub.



> 1\. Do _all_ other hubs support it? (usys / uselect) All equally well / reliably?

Yes.



> 2\. Which other hubs are compatible with the motors that come in Boost sets? (I'm considering 88009-1 and 88012-1)

Everything Powered Up is compatible with everything Powered Up. If the connector fits, it will work.

**Reply by tomsaffell** [reply]
That's great news - I like not having to buy a new hub. I see in the API docs:

`Reads one byte from standard input without blocking.`

That should work. What is the source of the standard input? Also a bluetooth keyboard? 
Are there any examples of using read_input_bytes that I can study? (I searched, but dont see any)

thanks again

**Reply by dlech** [reply]
> That should work. What is the source of the standard input?

The terminal in Pybricks Code.

> Also a bluetooth keyboard?

Only if the Bluetooth keyboard is connected to your computer.

> Are there any examples of using read_input_bytes that I can study? 

https://github.com/pybricks/pybricksdev/issues/64#issuecomment-1741844971

**Comment by tomsaffell** [comment]
Thanks for the pointer to comment #64. I had to tinker with it for a while, but I finally got it working, so leaving it here for the next person. If you see any possible ways to improve it please suggest them

```python

from pybricks.hubs import MoveHub
from pybricks.pupdevices import Motor, ColorDistanceSensor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch
from pybricks.tools import read_input_byte, run_task


hub = MoveHub()



async def read_line():
    print('read_line()')
    line = ''
    while True:
        b = read_input_byte()

        # no input was available yet
        if b is None:
           # give other tasks a chance to run
            await wait(0)
            # then check for another byte
            continue

        # ignore carriage return
        #if b == CR:
        #    continue

        # line feed indicates end of line
        if b == 13:
            return line

        line = line + chr(b)

async def red():
    hub.light.on(Color.RED)

async def green():
    hub.light.on(Color.GREEN)


async def main():
    print('main()')
    while True:

        # wait for a command from the PC
        command = await read_line()
        print('command: ' + str(command))
        if command == 'lr':
            await red()
        elif command == 'lg':
            await green()        
                

run_task(main())

```
